File: Java/data_structures/node_implementation/Queue.java
package Java.data_structures.node_implementation;

/**
 * Java Implementation of Queue using Nodes (LinkedList Queue)
 * Dynamic sizing, no capacity specification needed, simply a chain of pointers pointing to the next element
 * Init as -->  Queue<T> q = new Queue<>();
 * Methods -->  void push(T value);
 *              T peek();
 *              T poll();
 *              int size();
 *              boolean isEmpty();
 *              boolean contains();
 *              void clear();
 *              @Override boolean equals(Object obj);
 *              @Override int hashCode();
 *              @Override String toString();
 * @param <T> Generic Object Limiter
 * @author Nikhil Daehee Agarwal
 */
public class Queue<T> {

    private Node<T> front;
    private Node<T> end;
    private int size;

    /**
     * Default Constructor for Queue Object
     * Sets front and end pointers to null and size to 0
     */
    public Queue(){
        this.front = null;
        this.end = null;
        this.size = 0;
    }

    /**
     * push a T value to the back of the Queue
     * @param value Generic Object T
     */
    public void push(T value){
        if(front==null){
            end = new Node<>(value);
            front = end;
        }else{
            Node<T> adder = new Node<>(value);
            end.next(adder);
            end = end.next();
        }
        size++;
    }

    /**
     * Grabs the value at the front of the Queue
     * @return error if no such node found, T value if there exists a node at the front of the queue
     */
    public T peek(){
        if(front==null){
            throw new NullPointerException("Queue is empty!");
        }
        return front.val();
    }

    /**
     * Grabs the value and the front of the Queue and removes it from the list.
     * @return error if no such node found, T value if there exists a node at the front of the Queue.
     */
    public T poll(){
        if(front==null){
            throw new NullPointerException("Queue is empty!");
        }
        T valToReturn = front.val();
        front = front.next();
        if(front == null){
            end = null;
        }
        size--;
        return valToReturn;
    }

    /**
     * gets the size of our Queue Object (number of elements in the Queue)
     * @return int size
     */
    public int size(){
        return size;
    }

    public void clear(){
        this.size = 0;
        this.front = null;
        this.end = null;
    }

    /**
     * Checks to see if there are no elements in our Queue.
     * @return true if size is 0, else otherwise.
     */
    public boolean isEmpty(){
        return size == 0;
    }

    /**
     * Checks our Queue to see the values exists in the Queue.
     * @param value Generic Object we are searching for
     * @return ture if the value was found, false otherwise.
     */
    public boolean contains(T value){
        Node<T> start = front;
        while(start!=null){
            if(start.val().equals(value)){
                return true;
            }
            start = start.next();
        }
        return false;
    }

    /**
     * HashCode of our Queue based on the sum of all the unique hashCodes of the individual elements.
     * Curtails the hashcode to take care of overflow
     * @return int hashCOde
     */
    @Override
    public int hashCode(){
        long sum = 0;
        Node<T> start = front;
        while(start!=null){
            sum += start.val().hashCode();
            sum %= Integer.MAX_VALUE;
            start = start.next();
        }
        sum = sum < 0 ? sum * (-1) : sum;
        return (int) sum;
    }

    /**
     * Checks to see if two queues are equal
     * @param obj The Second queue that we want to compare
     * @return true if the two queues are the same size and contain all the same elements in order, false otherwise
     */
    @Override
    public boolean equals(Object obj){
        Queue<T> q = (Queue<T>) obj;
        if(q.size() != this.size()){
            return false;
        }
        Node<T> s1 = q.front;
        Node<T> s2 = this.front;
        while(s1!=null){
            if(!s1.val().equals(s2.val())){
                return false;
            }
            s1 = s1.next();
            s2 = s2.next();
        }
        return true;
    }

    /**
     * Returns the Queue as visualized string
     * @return String of the Queue
     */
    @Override
    public String toString(){
        String ans = "[";
        Node<T> start = front;
        while(start != null){
            ans += start.val();
            if (start.next()!=null) {
                ans += ", ";
            }
            start = start.next();
        }
        return "<-out- "+ans + "] <-in-";
    }

}

File: C/data_structures/node_implementation/queue.c
/**
 * Implementation of a Queue data structure in C.
 * Uses Node structures linked together into a singly linked list.
 * Queue follows First in First Out (FIFO) principle
 * Init as -->  Queue queue1;
 *              initQueue(&queue1);
 * Destroy -->  endQueue(&queue1);
 * Methods -->  void push(Queue* q, int value);
 *              void printQueue(Queue* q);
 *              int isEmpty(Queue* q);
 *              int size(Queue* q);
 *              int peek(Queue* q);
 *              int pop(Queue* q);
 *              int contains(Queue* q, int value);
 *              int count(Queue* q, int value);
 * @author Nikhil Daehee Agarwal
*/

#include <stdio.h>
#include <stdlib.h>

const int INIT_SIZE = 0;
const int EMPTY_QUEUE_ERROR = 7;

/**
 * structure definition for Node structure (Makes up the compoenets of the Linked List);
*/
struct Node{
    int val;
    struct Node* next;
};

/**
 * structure definition for Queue (holds size of linked List and pointers to the front and end of the list);
*/
typedef struct{
    int size;
    struct Node* front;
    struct Node* end;
} Queue;

/**
 * Initializes Queue structure
 * Sets queue size to 0 and pointers to NULL
*/
void initQueue(Queue* q){
    q->size = INIT_SIZE;
    q->front = NULL;
    q->end = NULL;
}

/**
 * Push an int into the Queue
 * Dynamically allocates memeory for new node structure and adds node structure to the end of our linked list.
*/
void push(Queue* q, int value){
    if(q->size == 0){
        q->front = (struct Node*)malloc(sizeof(struct Node));
        q->front->val = value;
        q->end = q->front;
        q->end->next = NULL;
    }else{
        struct Node* newNode = (struct Node*)malloc(sizeof(struct Node));
        newNode->val = value;
        newNode->next = NULL;
        q->end->next = newNode;
        q->end = q->end->next;
    }
    q->size++;
}

/**
 * Prints the queue as an observable in the terminal
*/
void printQueue(Queue* q){
    printf("out [");
    struct Node* start = q->front;
    while(start!=NULL){
        printf("%d",start->val);
        if(start->next != NULL){
            printf(", ");
        }
        start = start->next;
    }
    printf("] in\n");
}

/**
 * Method to destroy Queue structure and free all dynamically memory
*/
void endQueue(Queue* q){
    struct Node* start = q->front;
    while(start!=NULL){
        struct Node* toFree = start;
        start = start->next;
        free(toFree);
    }
}

/**
 * Returns 1 if the Queue is empty (size is 0), returns 0 if the queue still points to some nodes
*/
int isEmpty(Queue* q){
    if(q->size == 0){
        return 1;
    }
    return 0;
}

/**
 * returns the size of the Queue (the number of elements it holds)
*/
int size(Queue* q){
    return q->size;
}

/**
 * returns the value at front of the Queue, or prints error message if queue is empty
*/
int peek(Queue* q){
    if(q->size == 0){
        printf("Empty Queue Error");
        exit(EMPTY_QUEUE_ERROR);
    }
    return q->front->val;
}

/**
 * returns the value at the front of the queue, after it removes this value fron the Queue and
 * frees the memory that was dynamically allocated for it.
*/
int poll(Queue* q){
    if(q->size == 0){
        printf("Empty Queue Error");
        exit(EMPTY_QUEUE_ERROR);
    }
    q->size--;
    int value = q->front->val;
    struct Node* toFree = q->front;
    q->front = q->front->next;
    if(q->front == NULL){
        q->end = NULL;
    }
    free(toFree);
    return value;
}

/**
 * returns 1 if the queue contains a given value, 0 otherwise
*/
int contains(Queue* q, int value){
    struct Node* start = q->front;
    while(start!=NULL){
        if(start->val == value){
            return 1;
        }
        start = start->next;
    }
    return 0;
}

/**
 * returns the number of times a given value is found in the queue
*/
int count(Queue* q, int value){
    int ans = 0;
    struct Node* start = q->front;
    while(start!=NULL){
        if(start->val == value){
            ans++;
        }
        start = start->next;
    }
    return ans;
}